cpp/oneapi/dal/algo/decision_tree/common.hpp

```diff
@@ -0,0 +1,193 @@
+/// :capterm:`classification problem <classification>`.
+struct classification {};
+
+/// Tag-type that parameterizes entities used for solving
+/// :capterm:`regression problem <regression>`.
+struct regression {};
+
+/// Alias tag-type for classification task.
+using by_default = classification;
+} // namespace v1
+
+using v1::classification;
+using v1::regression;
+using v1::by_default;
+
+} // namespace task
+
+namespace detail {
+namespace v1 {
+
+template <typename Task>
+class node_info_impl;
+
+template <typename Task>
+class split_node_info_impl;
+
+template <typename Task>
+class leaf_node_info_impl;
+
+template <typename Task>
+constexpr bool is_valid_task_v =
+    dal::detail::is_one_of_v<Task, task::classification, task::regression>;
+
+} // namespace v1
+
+using v1::node_info_impl;
+using v1::split_node_info_impl;
+using v1::leaf_node_info_impl;
+using v1::is_valid_task_v;
+} // namespace detail
+
+namespace v1 {
+
+/// Class containing base node info in decision tree
+template <typename Task = task::by_default>
+class node_info : public base {
+    static_assert(detail::is_valid_task_v<Task>);
+    friend dal::detail::pimpl_accessor;
+
+    using task_t = Task;
+
+public:
+    node_info();
+    node_info(const node_info<task_t>&);
+    node_info<task_t>& operator=(const node_info<task_t>&);
+
+    /// Number of connections between the node and the root
+    std::int64_t get_level() const;
+    /// Measure of the homogeneity of the response variable at the node (i.e., the value of the criterion)
+    double get_impurity() const;
+    /// Number of samples at the node
+    std::int64_t get_sample_count() const;
+
+protected:
+    using impl_t = detail::node_info_impl<task_t>;
+
+    explicit node_info(impl_t* impl);
+    explicit node_info(const dal::detail::pimpl<impl_t>& impl);
+
+    dal::detail::pimpl<impl_t> impl_;
+};
+
+/// Class containing description of split node in decision tree
+template <typename Task = task::by_default>
+class split_node_info : public node_info<Task> {
+    static_assert(detail::is_valid_task_v<Task>);
+    friend dal::detail::pimpl_accessor;
+
+    using task_t = Task;
+    using impl_t = detail::split_node_info_impl<task_t>;
+
+public:
+    split_node_info();
+    split_node_info(const split_node_info<task_t>&);
+    split_node_info<task_t>& operator=(const split_node_info<task_t>&);
+
+    /// Feature used for splitting the node
+    std::int64_t get_feature_index() const;
+    /// Threshold value at the node
+    double get_feature_value() const;
+
+private:
+    explicit split_node_info(impl_t* impl);
+};
+
+template <typename Task>
+class leaf_node_info;
+
+/// Class containing description of leaf node in classification decision tree
+template <>
+class leaf_node_info<task::classification> : public node_info<task::classification> {
+    using task_t = task::classification;
+    using impl_t = detail::leaf_node_info_impl<task_t>;
+
+public:
+    leaf_node_info() = delete;
```

Comment:
Not required if there is non-default constructor declared

Reply:
removed